File: VelibStations_iOS/VelibStations/Location/LocationManager.swift
//
//  LocationManager.swift
//  VelibStations
//
//  Created by Safi on 01/10/2019.
//

import Foundation
import CoreLocation
import CommonKit

class LocationManager : AddOn, ILocationManager, CLLocationManagerDelegate {
    private var mLocation: CLLocation? = nil
    private var mLocationManger: CLLocationManager? = nil
    
    override init() {
        super.init()
    
        mLocationManger = CLLocationManager()
        mLocationManger?.delegate = self
        mLocationManger?.distanceFilter = kCLDistanceFilterNone
        mLocationManger?.desiredAccuracy = kCLLocationAccuracyBest
    }
    
    func requestUpdates() {
        cancelUpdates()
        
        let isPermission = (CLLocationManager.authorizationStatus() == CLAuthorizationStatus.authorizedWhenInUse) || (CLLocationManager.authorizationStatus() == CLAuthorizationStatus.authorizedAlways)
        
        if (isPermission) {
            changeLocationUpdates(enable: isPermission)
            let eventManager: IEventManager? = getAddOn(type: AddOnType().EVENT_MANAGER) as? IEventManager
            let event = Event(type: LocationEventType.REQUEST_UPDATES, data: nil, error: nil)
            eventManager?.send(event: event)
        } else {
            mLocationManger?.requestWhenInUseAuthorization()
        }
    }
    
    func cancelUpdates() {
        changeLocationUpdates(enable: false)
    }
    
    func getLocation() -> CLLocation? {
        return mLocation
    }
    
    func changeLocationUpdates(enable: Bool) {
        if (enable) {
            mLocationManger?.startUpdatingLocation()
        } else {
            mLocationManger?.stopUpdatingLocation()
        }
    }
    
    func locationManager(_ manager: CLLocationManager, didChangeAuthorization status: CLAuthorizationStatus) {
        let isPermission = (CLLocationManager.authorizationStatus() == CLAuthorizationStatus.authorizedWhenInUse) || (CLLocationManager.authorizationStatus() == CLAuthorizationStatus.authorizedAlways)
        changeLocationUpdates(enable: isPermission)
        let eventManager: IEventManager? = getAddOn(type: AddOnType().EVENT_MANAGER) as? IEventManager
        var error: Error? = nil
        if (!isPermission) {
            error = Error(code: "", message: "")
        }
        let event = Event(type: LocationEventType.REQUEST_UPDATES, data: nil, error: error)
        eventManager?.send(event: event)
    }
    
    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        if let location = locations.first {
            if (mLocation == nil || mLocation?.coordinate.latitude != location.coordinate.latitude || mLocation?.coordinate.longitude != location.coordinate.longitude) {
                mLocation = CLLocation(latitude: location.coordinate.latitude, longitude: location.coordinate.longitude)
                let eventManager: IEventManager? = getAddOn(type: AddOnType().EVENT_MANAGER) as? IEventManager
                let event = Event(type: LocationEventType.UPDATE_LOCATION, data: mLocation, error: nil)
                eventManager?.send(event: event)
            }
        }
    }
}

File: VelibStations_Android/common/src/iosMain/kotlin/com/muhammedsafiulazam/common/location/LocationManager.kt
package com.muhammedsafiulazam.common.location

import com.muhammedsafiulazam.common.addon.AddOn
import com.muhammedsafiulazam.common.addon.AddOnType
import com.muhammedsafiulazam.common.event.Event
import com.muhammedsafiulazam.common.event.IEventManager
import com.muhammedsafiulazam.common.location.event.LocationEventType
import com.muhammedsafiulazam.common.model.Location
import com.muhammedsafiulazam.common.service.model.Error
import kotlinx.cinterop.useContents
import platform.CoreLocation.*
import platform.UIKit.UIApplication
import platform.darwin.NSObject

actual class LocationManager : AddOn(), ILocationManager {
    private var mLocation: Location? = null
    private var mLocationManager: CLLocationManager? = null

    init {
        mLocationManager = CLLocationManager()
        mLocationManager?.delegate = object : NSObject(), CLLocationManagerDelegateProtocol {
            override fun locationManager(manager: CLLocationManager, didChangeAuthorizationStatus: CLAuthorizationStatus) {
                onLocationManager(manager, didChangeAuthorizationStatus)
            }

            override fun locationManager(manager: CLLocationManager, didUpdateLocations: List<*>) {
                onLocationManager(manager, didUpdateLocations)
            }
        }

        mLocationManager?.distanceFilter = kCLDistanceFilterNone
        mLocationManager?.desiredAccuracy = kCLLocationAccuracyBest
    }

    actual override fun requestUpdates() {
        cancelUpdates()

        val isPermission = (CLLocationManager.authorizationStatus() == kCLAuthorizationStatusAuthorizedWhenInUse) || (CLLocationManager.authorizationStatus() == kCLAuthorizationStatusAuthorizedAlways)

        if (isPermission) {
            changeLocationUpdates(isPermission)
            val eventManager: IEventManager? = getAddOn(AddOnType.EVENT_MANAGER) as IEventManager?
            val event = Event(LocationEventType.REQUEST_UPDATES, null, null)
            eventManager?.send(event)

        } else {
            mLocationManager?.requestWhenInUseAuthorization()
        }
    }
    actual override fun cancelUpdates() {
        changeLocationUpdates(false)

    }
    actual override fun getLocation() : Location? {
        return mLocation
    }

    private fun changeLocationUpdates(enable: Boolean) {
        if (enable) {
            mLocationManager?.startUpdatingLocation()
        } else {
            mLocationManager?.stopUpdatingLocation()
        }
    }

    private fun onLocationManager(manager: CLLocationManager, didChangeAuthorizationStatus: CLAuthorizationStatus) {
        val isPermission = (CLLocationManager.authorizationStatus() == kCLAuthorizationStatusAuthorizedWhenInUse) || (CLLocationManager.authorizationStatus() == kCLAuthorizationStatusAuthorizedAlways)
        changeLocationUpdates(isPermission)
        val eventManager: IEventManager? = getAddOn(AddOnType.EVENT_MANAGER) as IEventManager?
        var error: Error? = null
        if (!isPermission) {
            error = Error("", "")
        }
        val event = Event(LocationEventType.REQUEST_UPDATES, null, error)
        eventManager?.send(event)
    }

    private fun onLocationManager(manager: CLLocationManager, didUpdateLocations: List<*>) {
        (didUpdateLocations.last() as CLLocation).coordinate.useContents {
            if (mLocation == null || mLocation?.latitude != latitude || mLocation?.longitude != longitude) {
                mLocation = Location(latitude, longitude)
                val eventManager: IEventManager? = getAddOn(AddOnType.EVENT_MANAGER) as IEventManager?
                val event = Event(LocationEventType.UPDATE_LOCATION, mLocation, null)
                eventManager?.send(event)
            }
        }
    }
}